Add and Manage zones

sean-prz opened the pull request:
- fix:  #317 
- fix: #335 
## What

#### Users can now manage (add/refresh/delete) offline portions of the map called Zone. 
The informations about the zones, is downloaded on the local file storage, meaning the user can access this data, no matter what account, he's logged in, even as guest mode.


<img width="295" alt="image" src="https://github.com/user-attachments/assets/a497e8aa-b138-4ba3-83f1-5eb710de9117">
<img width="301" alt="image" src="https://github.com/user-attachments/assets/c84d15cc-1c96-44ca-af85-57f57156b761">



## How
- The `ZoneManagerScreen` load from files all zones and displays it, with possibility to refresh or delete it.
- The `ZoneSelectionScreen` allow user to add a new zone with code copied from `LocationPicker` and a dialog box to pick the name.
- The `ZoneSelectionScreen` write to the local storage the information about the zone.
- All the IOs are done in the `Zone.kt` file. Ensurint that the path and naming is standarised
- The `Zone` Object are stored as json object in` appDir/zones/${zone.uid}.json`

**Important** Currently when adding a zone, the tile and parking data is not downloaded, same for deletion and refresh. => Only the Zone Data Class is managed not the Parkings and not the Map. 
(Once we have implemented the function we should call them in the callbacks of the button, that add/refresh/delete.)

## Why
#### This is one of the first step in order to have an offline mode. We now have the interface to manage our offline plans. 


-- -
Overall: 80%
ui zone : 75%
model zone : 25%
<img width="151" alt="image" src="https://github.com/user-attachments/assets/ce6cd40d-19c1-4d1c-9ef8-d4a3b25d3d64">

matttsss commented on app/src/main/java/com/github/se/cyrcle/ui/zone/ZoneManagerScreen.kt:
```diff
@@ -23,19 +45,50 @@ import com.github.se.cyrcle.ui.theme.molecules.TopAppBar
+          // ===  === ===  ===
+
+          // === Content ===
+          Column(Modifier.fillMaxWidth()) {
```
Here I think a lazycolumn might be a better idea? We do not know the number of zones we have at compile time...

sean-prz replied:
Sure I can do a lazyColumn, did not think users would add more than 2 or 3 zones, but it's better

matttsss commented on app/src/main/java/com/github/se/cyrcle/ui/zone/ZoneManagerScreen.kt:
```diff
@@ -51,3 +104,57 @@ fun AddZoneButton(navigationActions: NavigationActions) {
 }
+// Display relevant information about a zone. Allows the user to refresh the zone or delete it.
+@Composable
+fun ZoneCard(zone: Zone, zones: MutableState<List<Zone>>) {
```
I guess that passing `zones` is ok, but wouldn't it be better to make it a `onDelete` callback, this might be more versatile?

sean-prz replied:
I don't like callback, I think they make the code harder to read 🙈 , especially here we don't need versatility as we would never call the `ZoneCard` with a different onDelete function

matttsss replied:
Honestly, I can understand... (specially when we look at the ViewModels xD)

Armetiis replied:
Honestly I also think it's a bit weird to pass the whole List to a card, but if you think that this way is better, it's ok

matttsss commented on app/src/main/java/com/github/se/cyrcle/ui/zone/ZoneSelectionScreen.kt:
```diff
@@ -1,26 +1,226 @@
+      mapViewModel.locationPickerState
+          .collectAsState() // state representing where the user is in the location selection
+  // process
+  mapViewModel
```
Would it not be better to move this function call into the `LaunchedEffect(Unit)` ? I'm not to familiar with this so I'm not sure if that would be better?

sean-prz replied:
It doesn't work, because it launch a new coroutine and the map is created before the camera is updated, making the call useless

matttsss reviewed (changes requested):
Overall a nice improvement... However I have a few issues with the UI:
* There are no size constraints on the name, is that a good idea?
* I also have a crash with nullptr exception, to reproduce:
    - Add zone
    - Try to add a second one
    (I have DM'ed you the crash info)

Could you try to fix this crash before we merge this PR?

sean-prz commented:
@matttsss 
I replaced the input text with the conditional one made by Maria (which I tweaked a bit) 
<img width="204" alt="image" src="https://github.com/user-attachments/assets/cdfe2dbf-5a26-482e-a898-4adbaa71f59e">
And fixed the crash.
The locationPickerState should have been reset in the `LaunchEffect(Unit) {...} `but somehow the `LaunchedEffect(locationPickerState) {...}` runs before and so made the app crash

matttsss reviewed (approved):
Nice, thank you for solving this issue!

Armetiis commented on app/src/main/java/com/github/se/cyrcle/model/zone/Zone.kt:
```diff
@@ -20,7 +25,101 @@ data class Zone(
+  ) : this(boundingBox, name, LocalDate.now(), generateZoneUid())
+
+  /** Returns a JSON representation of the zone. */
+  fun toJson(): String {
```
I feel like this function needs a lot of maintenance. As soon as you add a field you have to modify the toJson and fromJson functions. Wouldn't it be better to proceed like in ParkingRepositoryFirestore and have an adapter for the Bounding Box. Have a look at `BoundingBoxTypeAdapter()` provided by mapbox first to make sure it doesn't work how you'd like it to before looking into making a custom one.

If you prefer keeping a man made serializer, maybe you could use a json Builder as I really feel like this method is error prone

sean-prz replied:
@Armetiis , You're right it's not pretty.
I fixed it take a look at the corresponding commit
I had to make a customTypeAdapter for the localDate, and learn how to use Gson, but I think it looks cleaner now.

Armetiis commented on app/src/main/java/com/github/se/cyrcle/model/zone/Zone.kt:
```diff
@@ -20,7 +25,101 @@ data class Zone(
+      if (!zoneDir.exists()) {
+        zoneDir.mkdirs()
+      }
+      File(zoneDir, "${zone.uid}.json").writeText(zone.toJson())
```
beautiful

Armetiis commented on app/src/main/java/com/github/se/cyrcle/ui/addParking/location/LocationPickerBottomBar.kt:
I am a bit confused by the 138 lines changed. I guess this is just a ktfmt thing, but I couldn't identify the actual change. Plus you don't even use in your implementation in the end. So I guess it's nothing too important

Armetiis commented on app/src/main/java/com/github/se/cyrcle/ui/theme/atoms/Input.kt:
I think it's nice that you improve the atom, but make sure that you don't bring down the coverage. Adding branches in UI elements is the fastest way to divide coverage by 2

Armetiis commented on app/src/main/java/com/github/se/cyrcle/ui/zone/ZoneManagerScreen.kt:
```diff
@@ -46,8 +102,65 @@ fun AddZoneButton(navigationActions: NavigationActions) {
+            Icons.Filled.Refresh,
+            contentDescription = "Refresh",
+            modifier = Modifier.weight(0.2f).alpha(0.0f))
+        Icon(
```
Really smart to put invisible icon for alignment

Armetiis reviewed (approved):
Honestly very clean implementation, be it from a backend or UI perspective.
There's only the (de)serializer I am a bit worried about, but I don't think this change has to be addressed now. It can be addressed in the future, in particular if they need to evolve.

Keep up the good work. Soon we'll have a fully working offline mode 🫡

sean-prz commented on app/src/main/java/com/github/se/cyrcle/model/map/MapViewModel.kt:
```diff
@@ -142,6 +142,12 @@ class MapViewModel : ViewModel() {
     _cameraPosition.value = newCameraState
   }
 
+  fun unZoomActualCamera() {
```
fixed !

Armetiis reviewed (approved):
Thank you for addressing the changes! Everything seems to be still working perfect. Good job
